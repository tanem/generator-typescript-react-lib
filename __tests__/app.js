'use strict'

const fs = require('fs')
const glob = require('glob')
const helpers = require('yeoman-test')
const path = require('path')
const templateMap = require('../generators/app/template-map')

const getFiles = () =>
  glob
    .sync('**', {
      cwd: path.join(process.cwd(), 'generators/app/templates'),
      dot: true,
      nodir: true,
    })
    .map((file) => (templateMap.has(file) ? templateMap.get(file) : file))

const readFile = (file) => {
  const data = fs.readFileSync(file, 'utf-8')
  return path.extname(file) === '.json' ? JSON.parse(data) : data
}

const emailMock = jest.fn().mockResolvedValue('mock.user@mock.com')
const nameMock = jest.fn().mockResolvedValue('Mock Name')
const githubUsernameMock = jest.fn().mockResolvedValue('mockgithubusername')
const packageName = 'test-package'

beforeAll((done) => {
  helpers
    .run(path.join(__dirname, '../generators/app'))
    .withPrompts({
      author: 'John Smith <john.smith@js.com>',
      packageDescription: 'Test package.',
      packageName,
      umdGlobalName: 'TestPkg',
      githubUsername: 'jsmith',
    })
    .on('ready', (generator) => {
      generator.user.github.githubUsername = githubUsernameMock
      generator.user.git.name = nameMock
      generator.user.git.email = emailMock
    })
    .on('end', done)
})

test.each(getFiles())('creates %s', (file) => {
  if (path.basename(file) === 'package.json') {
    const data = readFile(file)
    const { dependencies, devDependencies, peerDependencies } = data
    const setAsymmetricMatchers = (object) => {
      for (const key in object) {
        if (key !== packageName) {
          object[key] = expect.any(String)
        }
      }
      return object
    }
    expect(data).toMatchSnapshot({
      ...(dependencies
        ? { dependencies: setAsymmetricMatchers(dependencies) }
        : {}),
      ...(devDependencies
        ? { devDependencies: setAsymmetricMatchers(devDependencies) }
        : {}),
      ...(peerDependencies
        ? { peerDependencies: setAsymmetricMatchers(peerDependencies) }
        : {}),
    })
  } else {
    expect(readFile(file)).toMatchSnapshot()
  }
})
