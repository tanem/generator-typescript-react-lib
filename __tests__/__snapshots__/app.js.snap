// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates .eslintignore 1`] = `
"compiled/
coverage/
dist/
node_modules/"
`;

exports[`creates .eslintrc.js 1`] = `
"const path = require('path')

module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: path.join(__dirname, 'tsconfig.eslint.json'),
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
    },
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  extends: [
    'plugin:react/recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
  ],
  plugins: ['react', 'react-hooks', '@typescript-eslint'],
  rules: {
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/explicit-member-accessibility': 'off',
    'react-hooks/exhaustive-deps': 'warn',
    'react-hooks/rules-of-hooks': 'error',
  },
  overrides: [
    {
      files: ['*.js'],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
      },
    },
  ],
}
"
`;

exports[`creates .github/workflows/ci.yml 1`] = `
"name: CI

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  ci:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: '14'
      - run: npm ci
      - run: npm test
      - uses: codecov/codecov-action@v3
"
`;

exports[`creates .gitignore 1`] = `
"*.log
.DS_Store
compiled
coverage
dist
node_modules"
`;

exports[`creates .nvmrc 1`] = `"lts/*"`;

exports[`creates .prettierignore 1`] = `
"compiled
coverage
dist"
`;

exports[`creates .prettierrc.json 1`] = `
{
  "semi": false,
  "singleQuote": true,
}
`;

exports[`creates LICENSE 1`] = `
"MIT License

Copyright (c) 2023 John Smith <john.smith@js.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"
`;

exports[`creates README.md 1`] = `
"# test-package

[![npm version][npmv-image]][npmv-url]
[![build status][build-image]][build-url]
[![coverage status][codecov-image]][codecov-url]
[![npm downloads][npmd-image]][npmd-url]

> Test package.

## Basic Usage

\`\`\`jsx
import React from 'react'
import { render } from 'react-dom'

render(, document.getElementById('root'))
\`\`\`

## Live Examples

- [Basic Usage](https://codesandbox.io/)
- [API Example](https://codesandbox.io/)
- [UMD Build (Development)](https://codesandbox.io/)
- [UMD Build (Production)](https://codesandbox.io/)

## API

**Props**

- \`foo\` - Something something.
- \`bar\` - _Optional_ Something something. Defaults to \`null\`.

**Example**

\`\`\`jsx

\`\`\`

## Installation

\`\`\`
$ npm install test-package --save
\`\`\`

There are also UMD builds available via [unpkg](https://unpkg.com/):

- https://unpkg.com/test-package/dist/test-package.umd.development.js
- https://unpkg.com/test-package/dist/test-package.umd.production.js

For the non-minified development version, make sure you have already included:

- [\`React\`](https://unpkg.com/react/umd/react.development.js)
- [\`ReactDOM\`](https://unpkg.com/react-dom/umd/react-dom.development.js)
- [\`PropTypes\`](https://unpkg.com/prop-types/prop-types.js)

For the minified production version, make sure you have already included:

- [\`React\`](https://unpkg.com/react/umd/react.production.min.js)
- [\`ReactDOM\`](https://unpkg.com/react-dom/umd/react-dom.production.min.js)

## License

MIT

[build-image]: https://img.shields.io/github/workflow/status/jsmith/test-package/CI?style=flat-square
[build-url]: https://github.com/jsmith/test-package/actions?query=workflow%3ACI
[codecov-image]: https://img.shields.io/codecov/c/github/jsmith/test-package.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/jsmith/test-package
[npmv-image]: https://img.shields.io/npm/v/test-package.svg?style=flat-square
[npmv-url]: https://www.npmjs.com/package/test-package
[npmd-image]: https://img.shields.io/npm/dm/test-package.svg?style=flat-square
[npmd-url]: https://www.npmjs.com/package/test-package
"
`;

exports[`creates codecov.yml 1`] = `
"comment: off
coverage:
  status:
    patch: off
"
`;

exports[`creates examples/basic-usage/.env 1`] = `"REACT_APP_DESCRIPTION=$npm_package_description"`;

exports[`creates examples/basic-usage/.gitignore 1`] = `
"/node_modules
"
`;

exports[`creates examples/basic-usage/README.md 1`] = `
"# TestPkg Basic Usage Example

This project was bootstrapped with [Create React App](https://github.com/facebookincubator/create-react-app).

## Available Scripts

In the project directory, you can run:

### \`npm start\`

Runs the app in the development mode.<br>
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.<br>
You will also see any lint errors in the console.
"
`;

exports[`creates examples/basic-usage/package.json 1`] = `
{
  "dependencies": {
    "react": Any<String>,
    "react-dom": Any<String>,
    "react-scripts": Any<String>,
    "test-package": "latest",
  },
  "description": "TestPkg Basic Usage Example",
  "keywords": [
    "test-package",
  ],
  "name": "basic-usage",
  "private": true,
  "scripts": {
    "start": "react-scripts start",
  },
  "version": "0.1.0",
}
`;

exports[`creates examples/basic-usage/public/index.html 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <title>%REACT_APP_DESCRIPTION%</title>
</head>

<body>
  <div id="root"></div>
</body>

</html>"
`;

exports[`creates examples/basic-usage/public/svg.svg 1`] = `
"<svg width="40" height="40" version="1.1" xmlns="http://www.w3.org/2000/svg">
  <rect x="5" y="5" width="30" height="30" stroke="black" fill="transparent" stroke-width="5" />
</svg>"
`;

exports[`creates examples/basic-usage/src/index.js 1`] = `
"import C from 'test-package'
import React from 'react'
import ReactDOM from 'react-dom'

ReactDOM.render(<C />, document.getElementById('root'))
"
`;

exports[`creates index.js 1`] = `
"'use strict'

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./test-package.cjs.production.js')
} else {
  module.exports = require('./test-package.cjs.development.js')
}
"
`;

exports[`creates package.json 1`] = `
{
  "author": "John Smith <john.smith@js.com>",
  "bugs": {
    "url": "github:jsmith/test-package/issues",
  },
  "dependencies": {
    "@babel/runtime": Any<String>,
    "prop-types": Any<String>,
    "tslib": Any<String>,
  },
  "description": "Test package.",
  "devDependencies": {
    "@babel/core": Any<String>,
    "@babel/plugin-proposal-class-properties": Any<String>,
    "@babel/plugin-transform-runtime": Any<String>,
    "@babel/preset-env": Any<String>,
    "@babel/preset-react": Any<String>,
    "@testing-library/jest-dom": Any<String>,
    "@testing-library/react": Any<String>,
    "@testing-library/user-event": Any<String>,
    "@types/jest": Any<String>,
    "@types/prop-types": Any<String>,
    "@types/react": Any<String>,
    "@types/react-dom": Any<String>,
    "@typescript-eslint/eslint-plugin": Any<String>,
    "@typescript-eslint/parser": Any<String>,
    "babel-core": Any<String>,
    "babel-plugin-transform-react-remove-prop-types": Any<String>,
    "cross-env": Any<String>,
    "eslint": Any<String>,
    "eslint-config-prettier": Any<String>,
    "eslint-plugin-react": Any<String>,
    "eslint-plugin-react-hooks": Any<String>,
    "jest": Any<String>,
    "npm-run-all": Any<String>,
    "prettier": Any<String>,
    "react": Any<String>,
    "react-dom": Any<String>,
    "react-test-renderer": Any<String>,
    "regenerator-runtime": Any<String>,
    "rollup": Any<String>,
    "rollup-plugin-babel": Any<String>,
    "rollup-plugin-commonjs": Any<String>,
    "rollup-plugin-filesize": Any<String>,
    "rollup-plugin-node-resolve": Any<String>,
    "rollup-plugin-replace": Any<String>,
    "rollup-plugin-size-snapshot": Any<String>,
    "rollup-plugin-sourcemaps": Any<String>,
    "rollup-plugin-terser": Any<String>,
    "shx": Any<String>,
    "tanem-scripts": Any<String>,
    "ts-jest": Any<String>,
    "typescript": Any<String>,
  },
  "files": [
    "dist",
  ],
  "homepage": "github:jsmith/test-package",
  "jsnext:main": "dist/test-package.esm.js",
  "keywords": [
    "javascript",
    "react",
    "typescript",
  ],
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/test-package.esm.js",
  "name": "test-package",
  "peerDependencies": {
    "react": Any<String>,
    "react-dom": Any<String>,
  },
  "repository": {
    "type": "git",
    "url": "github:jsmith/test-package",
  },
  "scripts": {
    "build": "run-s clean compile bundle",
    "bundle": "rollup -c",
    "check:format": "prettier --list-different "**/*.{js,ts,tsx}"",
    "check:types": "tsc --noEmit",
    "clean": "run-p clean:*",
    "clean:compiled": "shx rm -rf compiled",
    "clean:coverage": "shx rm -rf coverage",
    "clean:dist": "shx rm -rf dist",
    "compile": "tsc -p tsconfig.base.json",
    "format": "prettier --write "**/*.{js,ts,tsx}"",
    "lint": "eslint . --ext .js,.ts",
    "postbundle": "npm run clean:compiled && shx cp ./index.js ./dist/index.js",
    "release": "tanem-scripts release",
    "test": "run-s check:* lint build test:*",
    "test:cjs": "jest --config ./scripts/jest/config.cjs.js",
    "test:cjsprod": "cross-env NODE_ENV=production jest --config ./scripts/jest/config.cjs.js",
    "test:es": "jest --config ./scripts/jest/config.es.js",
    "test:src": "jest --config ./scripts/jest/config.src.js",
    "test:umd": "jest --config ./scripts/jest/config.umd.js",
    "test:umdprod": "jest --config ./scripts/jest/config.umdprod.js",
  },
  "typings": "dist/index.d.ts",
  "version": "0.0.1",
}
`;

exports[`creates renovate.json 1`] = `
{
  "extends": [
    "config:js-lib",
    ":rebaseStalePrs",
    ":automergeMinor",
    ":label(internal)",
    ":enableVulnerabilityAlerts",
  ],
  "ignorePaths": [
    "**/node_modules/**",
  ],
  "lockFileMaintenance": {
    "enabled": true,
  },
  "packageRules": [
    {
      "sourceUrlPrefixes": [
        "https://github.com/tanem/",
      ],
      "stabilityDays": 0,
    },
    {
      "depTypeList": [
        "dependencies",
      ],
      "rangeStrategy": "bump",
    },
  ],
  "stabilityDays": 3,
}
`;

exports[`creates rollup.config.js 1`] = `
"import babel from 'rollup-plugin-babel'
import commonjs from 'rollup-plugin-commonjs'
import nodeResolve from 'rollup-plugin-node-resolve'
import replace from 'rollup-plugin-replace'
import { sizeSnapshot } from 'rollup-plugin-size-snapshot'
import sourcemaps from 'rollup-plugin-sourcemaps'
import { terser } from 'rollup-plugin-terser'
import pkg from './package.json'

const CJS_DEV = 'CJS_DEV'
const CJS_PROD = 'CJS_PROD'
const ES = 'ES'
const UMD_DEV = 'UMD_DEV'
const UMD_PROD = 'UMD_PROD'

const input = './compiled/index.js'

const getGlobals = (bundleType) => {
  const baseGlobals = {
    'react-dom': 'ReactDOM',
    react: 'React',
  }

  switch (bundleType) {
    case UMD_DEV:
      return { ...baseGlobals, 'prop-types': 'PropTypes' }
    case UMD_PROD:
      return baseGlobals
    default:
      return {}
  }
}

const getExternal = (bundleType) => {
  const peerDependencies = Object.keys(pkg.peerDependencies)
  const dependencies = Object.keys(pkg.dependencies)

  // Hat-tip: https://github.com/rollup/rollup-plugin-babel/issues/148#issuecomment-399696316.
  const makeExternalPredicate = (externals) => {
    if (externals.length === 0) {
      return () => false
    }
    const pattern = new RegExp(\`^(\${externals.join('|')})($|/)\`)
    return (id) => pattern.test(id)
  }

  switch (bundleType) {
    case CJS_DEV:
    case CJS_PROD:
    case ES:
      return makeExternalPredicate([...peerDependencies, ...dependencies])
    case UMD_DEV:
      return makeExternalPredicate([...peerDependencies, 'prop-types'])
    default:
      return makeExternalPredicate(peerDependencies)
  }
}

const isProduction = (bundleType) =>
  bundleType === CJS_PROD || bundleType === UMD_PROD

const getBabelConfig = (bundleType) => {
  const options = {
    babelrc: false,
    exclude: 'node_modules/**',
    presets: [['@babel/env', { loose: true, modules: false }], '@babel/react'],
    plugins: ['@babel/transform-runtime'],
    runtimeHelpers: true,
  }

  switch (bundleType) {
    case ES:
      return {
        ...options,
        plugins: [
          ...options.plugins,
          ['transform-react-remove-prop-types', { mode: 'wrap' }],
        ],
      }
    case UMD_PROD:
    case CJS_PROD:
      return {
        ...options,
        plugins: [
          ...options.plugins,
          ['transform-react-remove-prop-types', { removeImport: true }],
        ],
      }
    default:
      return options
  }
}

const getPlugins = (bundleType) => [
  nodeResolve(),
  commonjs({
    include: 'node_modules/**',
    namedExports: {
      'node_modules/prop-types/index.js': [
        'any',
        'array',
        'arrayOf',
        'bool',
        'element',
        'exact',
        'func',
        'instanceOf',
        'node',
        'number',
        'object',
        'objectOf',
        'oneOf',
        'oneOfType',
        'shape',
        'string',
        'symbol',
      ],
    },
  }),
  babel(getBabelConfig(bundleType)),
  replace({
    'process.env.NODE_ENV': JSON.stringify(
      isProduction(bundleType) ? 'production' : 'development',
    ),
  }),
  sourcemaps(),
  sizeSnapshot(),
  isProduction(bundleType) &&
    terser({
      output: { comments: false },
      compress: {
        keep_infinity: true,
        pure_getters: true,
      },
      warnings: true,
      ecma: 5,
      toplevel: false,
    }),
]

const getCjsConfig = (bundleType) => ({
  input,
  external: getExternal(bundleType),
  output: {
    exports: 'default',
    file: \`dist/test-package.cjs.\${
      isProduction(bundleType) ? 'production' : 'development'
    }.js\`,
    format: 'cjs',
    sourcemap: true,
  },
  plugins: getPlugins(bundleType),
})

const getEsConfig = () => ({
  input,
  external: getExternal(ES),
  output: {
    file: pkg.module,
    format: 'es',
    sourcemap: true,
  },
  plugins: getPlugins(ES),
})

const getUmdConfig = (bundleType) => ({
  input,
  external: getExternal(bundleType),
  output: {
    file: \`dist/test-package.umd.\${
      isProduction(bundleType) ? 'production' : 'development'
    }.js\`,
    format: 'umd',
    globals: getGlobals(bundleType),
    name: 'TestPkg',
    sourcemap: true,
  },
  plugins: getPlugins(bundleType),
})

export default [
  getCjsConfig(CJS_DEV),
  getCjsConfig(CJS_PROD),
  getEsConfig(),
  getUmdConfig(UMD_DEV),
  getUmdConfig(UMD_PROD),
]
"
`;

exports[`creates scripts/jest/config.cjs.js 1`] = `
"const srcConfig = require('./config.src')

module.exports = Object.assign({}, srcConfig, {
  collectCoverage: false,
  moduleNameMapper: {
    '^../src$': \`<rootDir>/dist/index.js\`,
  },
})
"
`;

exports[`creates scripts/jest/config.es.js 1`] = `
"const srcConfig = require('./config.src')

module.exports = Object.assign({}, srcConfig, {
  collectCoverage: false,
  moduleNameMapper: {
    '^../src$': \`<rootDir>/dist/test-package.esm.js\`,
  },
})
"
`;

exports[`creates scripts/jest/config.src.js 1`] = `
"module.exports = {
  collectCoverage: true,
  collectCoverageFrom: ['src/*.{ts,tsx}'],
  moduleFileExtensions: ['ts', 'tsx', 'js', 'json'],
  preset: 'ts-jest',
  rootDir: process.cwd(),
  roots: ['<rootDir>/test'],
  setupFilesAfterEnv: ['<rootDir>/scripts/jest/setupJest.ts'],
  testEnvironment: 'jsdom',
  testMatch: ['<rootDir>/test/*.spec.ts?(x)'],
  transform: { '^.+\\\\.(js|tsx?)$': 'ts-jest' },
}
"
`;

exports[`creates scripts/jest/config.umd.js 1`] = `
"const srcConfig = require('./config.src')

module.exports = Object.assign({}, srcConfig, {
  collectCoverage: false,
  moduleNameMapper: {
    '^../src$': \`<rootDir>/dist/test-package.umd.development.js\`,
  },
})
"
`;

exports[`creates scripts/jest/config.umdprod.js 1`] = `
"const srcConfig = require('./config.src')

module.exports = Object.assign({}, srcConfig, {
  collectCoverage: false,
  moduleNameMapper: {
    '^../src$': \`<rootDir>/dist/test-package.umd.production.js\`,
  },
})
"
`;

exports[`creates scripts/jest/setupJest.ts 1`] = `
"import '@testing-library/jest-dom'
"
`;

exports[`creates src/index.tsx 1`] = `
"import PropTypes from 'prop-types'
import React from 'react'

interface Props {
  text: string
}

const C: React.FC<Props> = ({ text }) => <p>{text}</p>

C.propTypes = {
  text: PropTypes.string.isRequired,
}

export default C
"
`;

exports[`creates test/index.spec.tsx 1`] = `
"import { render, screen } from '@testing-library/react'
import React from 'react'
import C from '../src'

test('C', () => {
  const text = 'text'
  render(<C text={text} />)
  expect(screen.getByText(text)).toBeVisible()
})
"
`;

exports[`creates tsconfig.base.json 1`] = `
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "declaration": true,
    "declarationDir": "dist",
    "esModuleInterop": true,
    "importHelpers": true,
    "jsx": "react",
    "module": "esnext",
    "moduleResolution": "node",
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "compiled",
    "pretty": true,
    "removeComments": true,
    "sourceMap": true,
    "strict": true,
    "target": "es5",
  },
  "exclude": [
    "node_modules",
    "dist",
  ],
  "include": [
    "src",
  ],
}
`;

exports[`creates tsconfig.eslint.json 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
  },
  "extends": "./tsconfig.json",
  "files": [
    ".eslintrc.js",
    "index.js",
    "rollup.config.js",
  ],
  "include": [
    "examples",
    "scripts",
    "src",
    "test",
  ],
}
`;

exports[`creates tsconfig.json 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "declaration": false,
    "declarationDir": null,
    "module": "commonjs",
  },
  "exclude": [
    "node_modules",
  ],
  "extends": "./tsconfig.base.json",
  "include": [
    "scripts",
    "src",
    "test",
  ],
}
`;
